File: sight/native/share/pool.c
/*
 *  SIGHT - System information gathering hybrid tool
 *
 *  Copyright(c) 2007 Red Hat Middleware, LLC,
 *  and individual contributors as indicated by the @authors tag.
 *  See the copyright.txt in the distribution for a
 *  full listing of individual contributors.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library in the file COPYING.LIB;
 *  if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 * @author Mladen Turk
 *
 */

#include "sight.h"
#include "sight_local.h"
#include "sight_types.h"

static apr_status_t callback_object_cleanup(void *data)
{
    apr_status_t rv = APR_SUCCESS;
    sight_callback_t *cb = (sight_callback_t *)data;

    if (data) {
        JNIEnv *_E = NULL;
        jint em = sight_get_jnienv(&_E);
        if (_E && cb->object) {
            /* TODO: See if wee need to detach in case
             * callback call native methods
             */
            if (cb->method) {
                rv = (*_E)->CallIntMethod(_E, cb->object, cb->method, NULL);
                if ((*_E)->ExceptionCheck(_E)) {
                    /* Just clear any pending exceptions */
                    (*_E)->ExceptionClear(_E);
                }
            }
            (*_E)->DeleteGlobalRef(_E, cb->object);
        }
        sight_clr_jnienv(em);
        free(cb);
    }
    return rv;
}

SIGHT_EXPORT_DECLARE(jlong, Pool, register0)(SIGHT_STDARGS,
                                             jlong instance,
                                             jobject obj)
{
    sight_object_t *no = J2P(instance, sight_object_t *);
    sight_callback_t *cb = NULL;
    jclass c;

    UNREFERENCED_O;
    if (!no || !no->pool) {
        throwNullPointerException(_E, THROW_FMARK,
                                  sight_strerror(SIGHT_ENOPOOL));
        return 0;
    }
    if (!(cb = (sight_callback_t *)sight_calloc(_E,
                                    sizeof(sight_callback_t),
                                    THROW_FMARK))) {
        return 0;
    }
    SIGHT_LOCAL_TRY(no) {
        c = (*_E)->GetObjectClass(_E, obj);
        cb->name   = "cleanup";
        cb->msig   = "()I";
        cb->object = (*_E)->NewGlobalRef(_E, obj);
        cb->method = (*_E)->GetMethodID(_E, c, cb->name, cb->msig);

        apr_pool_cleanup_register(no->pool, (const void *)cb,
                                  callback_object_cleanup,
                                  apr_pool_cleanup_null);
    } SIGHT_LOCAL_END(no);
    return P2J(cb);
}

SIGHT_EXPORT_DECLARE(void, Pool, kill0)(SIGHT_STDARGS,
                                        jlong instance,
                                        jlong data)
{
    sight_object_t *no   = J2P(instance, sight_object_t *);
    sight_callback_t *cb = J2P(data, sight_callback_t *);

    UNREFERENCED_O;
    if (!no || !no->pool) {
        throwNullPointerException(_E, THROW_FMARK,
                                  sight_strerror(SIGHT_ENOPOOL));
        return;
    }
    if (!cb) {
        throwNullPointerException(_E, THROW_FMARK,
                                  sight_strerror(SIGHT_ENULL));
        return;
    }
    SIGHT_LOCAL_TRY(no) {
        apr_pool_cleanup_kill(no->pool, cb, callback_object_cleanup);
    } SIGHT_LOCAL_END(no);
    if (cb->object)
        (*_E)->DeleteGlobalRef(_E, cb->object);
    free(cb);
}

File: sight/java/org/jboss/sight/Pool.java
/*
 *  SIGHT - System information gathering hybrid tool
 *
 *  Copyright(c) 2007 Red Hat Middleware, LLC,
 *  and individual contributors as indicated by the @authors tag.
 *  See the copyright.txt in the distribution for a
 *  full listing of individual contributors.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library in the file COPYING.LIB;
 *  if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 */

package org.jboss.sight;

/** Pool
 *
 * @author Mladen Turk
 *
 */

public class Pool extends NativeObject
{

    private static native long  register0(long instance, IPoolCallback o)
                                    throws NullPointerException;
    private static native void  kill0(long instance, long data)
                                    throws NullPointerException;
    private static native void  notes0(long instance, long proc, int how);

    /**
     * Create new Pool with root as parent pool.
     */
    public Pool()
    {
        super(0, true);
    }

    /**
     * Create new NativeObject
     * @param parent The parent pool. If this is 0, the new pool is a root
     *        pool. If it is non-zero, the new pool will inherit all
     *        of its parent pool's attributes, except the apr_pool_t will
     *        be a sub-pool.
     */
    public Pool(Pool parent)
    {
        super(parent.POOL, true);
    }

    protected void onDestroy()
    {
        // Nothing
    }

    /**
     * Get native APR pool pointer.
     */
    public long getPoolHandle()
    {
        return POOL;
    }

    /**
     *  Create a new pool as sub-pool of this one.
     */
    public Pool create()
    {
        return new Pool(this);
    }

    /*
     * Cleanup
     *
     * Cleanups are performed in the reverse order they were registered.  That is:
     * Last In, First Out.  A cleanup function can safely allocate memory from
     * the pool that is being cleaned up. It can also safely register additional
     * cleanups which will be run LIFO, directly after the current cleanup
     * terminates.  Cleanups have to take caution in calling functions that
     * create subpools. Subpools, created during cleanup will NOT automatically
     * be cleaned up.  In other words, cleanups are to clean up after themselves.
     */

    /**
     * Register a function to be called when a pool is cleared or destroyed
     * @param callback The object to call when the pool is cleared
     *                 or destroyed.
     * @return The cleanup handler.
     */
    public long cleanupRegister(IPoolCallback callback)
        throws NullPointerException
    {
        return register0(INSTANCE, callback);
    }

    /**
     * Remove a previously registered cleanup function
     * @param data The cleanup handler to remove from cleanup
     */
    public void cleanupKill(long data)
        throws NullPointerException
    {
        kill0(INSTANCE, data);
    }

    /**
     * Register a process to be killed when a pool dies.
     * @param proc The Process to register
     * @param how How to kill the process, one of:
     * <PRE>
     *         KILL_NEVER         -- process is never sent any signals
     *         KILL_ALWAYS        -- process is sent SIGKILL on apr_pool_t cleanup
     *         KILL_AFTER_TIMEOUT -- SIGTERM, wait 3 seconds, SIGKILL
     *         JUST_WAIT          -- wait forever for the process to complete
     *         KILL_ONLY_ONCE     -- send SIGTERM and then wait
     * </PRE>
     */
    public void noteSubprocess(Process proc, KillConditions how)
    {
        notes0(INSTANCE, proc.INSTANCE, how.valueOf());
    }

}
